### DISCUSSION TITLE
MTestCurrentState.copy produces a shallow copy?

### DISCUSSION THREAD

**Discussion by mhpierre** [thelfer/tfel#564] [Q&A]
Hi Thomas,
I'm trying to get a deep copy of the `mtest` state in the `python` module. I create a state `s`, initialize it, copy it (`backup = s.copy()`), and run some calculations on the state `s`. I'm expecting the backup to hold the values of the initial states, but to the contrary it is as if the copy is shallow. I made a MWE: [MWE.zip](https://github.com/thelfer/tfel/files/15378770/MWE.zip). I was expecting `backup` to have the same value that I pulled from the state `s` as it was being copied.
Thanks in advance,

Maxime

**Comment by thelfer** [comment]
Hi @LordKthulhu,
You are right. It seems that I need to implement the `__copy__` method in in my `boost/python` bindings to work. Thus, I don't know what the default behaviour is. I never had to do this. Please open an issue if required.
In the meantime, you can indeed copy the members, such as `iv0`, `s0`, etc..

```python
backup_iv0 = s.iv0.copy()
```

**Reply by mhpierre** [reply]
Thanks for your answer, yes, it seems an implementations of `__copy__`and/or `__deepcopy__` is required as the object is not picklable. 
I was hoping to be able to do a retro-control on loads from internal state variables which really requires a deep copy, so I will create an issue.
